### PR TITLE
HDR Listener & HDR State Trigger

### LINKED ISSUE #1257: Would be nice to have Actions triggered by enabling/disabling HDR.
### Rules

- [X] I made myself familiar with the README.
- [X] I checked closed issues for similar ones and did not find a similar feature request mentioned before.
- [X] I understand that, if insufficient information was provided, my issue will be closed without an answer.

### Version

2.21.1

### OS

Windows 23H2 build 22631.3447

### Device

Legion 7i 16IAX7 (7th gen)

### Is your feature request related to a problem?

No, this is feature suggestion, merely a nice thing to have.

### How would you like the problem to be solved?

I would like to have an Actions trigger for enabling/disabling HDR. I was hoping to have an Actions configured so that screen brightness is automatically increased to 100%.

This is beneficial because HDR10 tone mapping scheme is mathematically designed for when display is operating at maximal brightness, and when it is not the tonemapping would be completely incorrect, making colors all wrong on top of incorrect brightness for different parts of the image. And this is by design as peak brightness is mathematically a prerequisite of dynamic range, so using HDR tonemapping when the display is not at its maximal designed brightness wouldn't be meaningful in the first place. This is also why typically when HDR is enabled on a dedicated external monitor, the monitor would automatically lock brightness to maximal brightness and disable brightness control. However Lenovo decided not to implement this on the Legion laptops that are capable of impressive 570nit of peak brightness for some seriously good HDR performance for a HDR400 rated screen, since after all this is a laptop and not a desktop-only external monitor. But it would still be really nice to have the option to have such a feature implemented when someone wants to.

### What alternatives have you considered?

Alternatively, instead of an Actions trigger for a user to implement this on their own using other Actions, this could also be provided as an option (like a toggle) for the existing HDR toggle on Dashboard page, although I'm not sure how well that would work if one toggles HDR using the Windows built-in Win+Alt+B shortcut. Assuming the shortcut works perfectly this does make sense, as there is simply no situation where it makes sense to have HDR enabled without also setting brightness to max, as long as this is provided as an option.

Currently we also have the alternative of using the "When game is running" trigger in Actions, however this is not a substitute, because HDR has uses outside gaming too, such as watching HDR encoded movies, doing professional photography/artistic/scientific work that requires an HDR display, etc., for which a trigger based on whether a game is running would be inadequate.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Ace-Radom**
Should close #1257.

Add a new `HDRListener` to watch HDR state changes, and two triggers `HDROnAutomationPipelineTrigger` & `HDROffAutomationPipelineTrigger`.

I set it to draft cuz there's still a problem needs to be solved. When I turn on HDR and then close the Lid (with an external monitor connected), there's about 50% chance that LLT crashes. Same thing happens when I open the Lid after that. It's most likely because of getting `Display` from `InternalDisplay` class, ~~but I'm not sure yet.~~

Bug fix status update: it should be fixed in [f441693](https://github.com/BartoszCichecki/LenovoLegionToolkit/pull/1262/commits/f441693b3debb96edd6834fb6c401aefa8cdefb1), at least it didn't crash in the past four / five times on my device. Multi-devices test done.

## Summary by CodeRabbit

- **New Features**
  - Introduced HDR control features, enabling automation triggers based on HDR settings.
  - Added support for HDR automation triggers in the user interface with localized display names.

- **Bug Fixes**
  - Improved equality check in a specific trigger for accurate behavior.

- **Documentation**
  - Updated documentation with new environment variables indicating HDR status.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -328,6 +328,12 @@ LLT automatically adds some variables to the process environment that can be acc
 - When external display is disconnected
 	- `LLT_IS_EXTERNAL_DISPLAY_CONNECTED=FALSE`
 
+- When HDR is on
+	- `LLT_IS_HDR_ON=TRUE`
+
+- When HDR is off
+	- `LLT_IS_HDR_OF=FALSE`
```
[the remark this answers is not in the document]
Comment (Ace-Radom) [reply README.md:335]: ahah, thx

File: LenovoLegionToolkit.Lib/Listeners/HDRListener.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+        return Task.CompletedTask;
+    }
+    
+    private void SystemEvents_DisplaySettingsChanged(object? sender, EventArgs e)
+    {
+        if (Log.Instance.IsTraceEnabled)
+            Log.Instance.Trace($"Event received.");
+
+        Task.Delay(100).ContinueWith(_ =>
```
Comment (Ace-Radom) [inline LenovoLegionToolkit.Lib/Listeners/HDRListener.cs:62]: In this architecture both `HDRListener` and `DisplayConfigurationListener` are watching `SystemEvents.DisplaySettingsChanged` signal. The thing (which caused the unreliability before) is, `HDRListener` needs `InternalDisplay` class to get HDR state, but `DisplayConfigurationListener` calls `InternalDisplay` class to refresh, and they run in parallel. Normally, `HDRListener` gains the lock in `InternalDisplay` first and get what it needs, and then `InternalDisplay` will be set to refresh-needed, which causes the display api `HDRListener` got becomes null. Therefore I added this delay here, in order to let `InternalDisplay` set its state first.

**Comment by Ace-Radom** [comment]
@BartoszCichecki should be done, ready for review

**Comment by Ace-Radom** [comment]
@BartoszCichecki new commits merged. Would you plz review this PR in your spare time? (Only a reminder xD)

**Comment by BartoszCichecki** [comment]
Hey @Ace-Radom I haven't forgotten! The PR looks good, but I have been busy with Gen 9 stuff recently. I will merge this when I am done with that.

**Comment by Ace-Radom** [comment]
> Hey @Ace-Radom I haven't forgotten! The PR looks good, but I have been busy with Gen 9 stuff recently. I will merge this when I am done with that.

Thx for telling me this xD. Hope you will finish it soon.

**Pull request merged**